### ISSUE TITLE
TERM and NO_COLOR envirnoment vars still not respected

### ISSUE THREAD

**Issue by ghost** [PowerShell/PowerShell#21160]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

As per "Enable use of TERM and NO_COLOR env var plaintext #14969" the TERM and NO_COLOR variables should disable coloring of errors in pwsh. But executing the showbug.bat in command.exe in the root directory of pwsh version 7 4 1 with this content:

    pwsh.exe -ExecutionPolicy Unrestricted ^
       -Command "$host.version ; " ^
       "Get-Item  Env:TERM ; " ^
       "Get-Item  Env:NO_COLOR ; " ^
       "Get-It1"

Shows:

     Major  Minor  Build  Revision
     -----  -----  -----  --------
     7      4      1      -1
     
     PSPath        : Microsoft.PowerShell.Core\Environment::TERM
     PSDrive       : Env
     PSProvider    : Microsoft.PowerShell.Core\Environment
     PSIsContainer : False
     Key           : TERM
     Value         : xterm-mono
     Name          : TERM
     
     
     PSPath        : Microsoft.PowerShell.Core\Environment::NO_COLOR
     PSDrive       : Env
     PSProvider    : Microsoft.PowerShell.Core\Environment
     PSIsContainer : False
     Key           : NO_COLOR
     Value         : True
     Name          : NO_COLOR
     
     Get-It1: The term 'Get-It1' is not recognized as a name of a cmdlet, function, script file, or executable program.
     Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
     
Here, the error message, triggered intentionally to demonstrate the bug, is RED colored.

### Expected behavior

```console
The error message should be printed without the colors.
```

### Actual behavior

```console
The error is RED colored, in spite of the environment variables requesting no colors and being clearly readable by the pwsh.exe.
```

### Error details

```console
The error in the above example is triggered intentionally to show that the error message is colored RED.
The error here is that the error message itself is displayed using colors, which is not what is supposed to happen according to the #14969
```

### Environment data

```powershell
PS C:\tmp\pwsh741> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.4.1
PSEdition                      Core
GitCommitId                    7.4.1
OS                             Microsoft Windows 6.3.9600
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

![pwsh74bug](https://github.com/PowerShell/PowerShell/assets/116484385/1e990e89-1aff-43f9-87e5-b7359b449df7)

**Comment by rhubarb-geek-nz** [comment]
NO_COLOR does get recognized and affects $PSStyle.OutputRending

```
$ NO_COLOR=true pwsh -c '$PSStyle.OutputRendering'
PlainText
$ pwsh -c '$PSStyle.OutputRendering'
Host
```

According to [about ANSI terminals](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_ansi_terminals?view=powershell-7.4)

> OutputRendering only applies to rendering in the Host, Out-File, and Out-String. Output from native executables isn't affected.

So I question if the error channel is included in that handling.

**Comment by janko-js** [comment]
The rationale is simple: if the colors are enforced by default, they will make the output for some users completely unreadable. There should be a way for users set a variable to prevent colors from being used:

https://no-color.org/

At the moment not only the error output is always colored, the same happens for the PS prompt input: on a gray background the input is yellow or light green both of which are unreadable.

NO_COLOR is not recognized and not respected for errors and for prompts.

**Comment by rhubarb-geek-nz** [comment]
> the input is yellow or light green both of which are unreadable.

Agreed!

[PowerShell on osx how to force to be completely monochrome](https://github.com/PowerShell/PowerShell/issues/21066)

**Comment by 237dmitry** [comment]
You can set monochrome pwsh:

![ss](https://github.com/PowerShell/PowerShell/assets/78153320/011d38ea-069e-49cb-8e1a-4637dc69b94d)

**Comment by 237dmitry** [comment]
> It means not using any colour escape codes in any output.

NO_COLOR also set colors with escape sequences, just reduce base 16 console colors to 2

**Comment by rhubarb-geek-nz** [comment]
escape 37 m sets the foreground colour to white, it does not mean 'no colour'. So if the background is white or yellow it becomes hard to read.

**Comment by 237dmitry** [comment]
> escape 37 m sets the foreground colour to white

No! This is 8th color of terminal color palette.

**Comment by rhubarb-geek-nz** [comment]
> > It means not using any colour escape codes in any output.
> 
> NO_COLOR also set colors with escape sequences, just reduce base 16 console colors to 2

From no-color.org

```Command-line software which adds ANSI color to its output by default should check for a NO_COLOR environment variable that, when present and not an empty string (regardless of its value), prevents the addition of ANSI color.
```

The goal is not have _any_ ANSI colour code escape sequences in the output.

**Comment by 237dmitry** [comment]
> prevents the addition of ANSI color.

How when terminal does understand what color it has to output? Any symbol, any color and any hotkey are esc-sequences

**Comment by rhubarb-geek-nz** [comment]
> How when terminal does understand what color it has to output? Any symbol, any color and any hotkey are esc-sequences

A true monochrome terminal does not have that problem, there are no colours to choose from, It may be white on black, green on black or amber on black, some monochrome terminals allow you to have reversed characters. In the same manner it is completely legitimate to have black on white. The application does not deal with any colours at all, it just prints text and it is printed how the terminal sees fit.

**Comment by 237dmitry** [comment]
> A true monochrome terminal does not have that problem

What does it mean "true"? Kernel console 80x25?

**Comment by rhubarb-geek-nz** [comment]
> > A true monochrome terminal does not have that problem
> 
> What does it mean "true"? Kernel console 80x25?

VT102, VT52 were all monochrome terminals

**Comment by 237dmitry** [comment]
> VT102, VT52 were all monochrome terminals

These are hardware terminals, let's get closer to life.

> The original xterm was monochrome as not everyone could afford colour.

I do not remember "original" xterm. But I do remember that xterm configured with ~/.Xresources

https://invisible-island.net/xterm/ctlseqs/ctlseqs.html

**Comment by rhubarb-geek-nz** [comment]
> These are hardware terminals, let's get closer to life.

Modern Macintosh Terminal Console offers many themes, including black on yellow. But PowerShell uses Yellow on Yellow, It also offers black and white. PowerShell gives you yellow on white. The terminals also work on according to the light/dark mode and will change colours during the day. So more that 50% of the time the 
PowerShell colour configuration is wrong, if you reconfigure it, then it will be wrong again when the dark/light mode changes. It will be wrong if you have multiple terminals open using different themes. So the solution is we want, when asked, is for PowerShell to _not do any colours at all_.

**Comment by 237dmitry** [comment]
> But PowerShell uses Yellow on Yellow

Just test two commands, there is difference in order of colors:

```
0..8 | % { Write-Host $([consoleColor] $_) -f $_ }
30..37 | % { "`e[$($_)mTEST" 
```

The .Net...

**Comment by janko-js** [comment]
> How when terminal does understand what color it has to output? Any symbol, any color and any hotkey are esc-sequences

For a programmer, the correct answer is to see how it's implemented elsewhere.

And example of using https://github.com/termcolor/termcolor 

![demo](https://github.com/PowerShell/PowerShell/assets/116484385/2df1117c-ce3f-4bb3-b6a3-1896a9863dcc)

[demo-temcolor-nocolor.zip](https://github.com/PowerShell/PowerShell/files/14098979/demo-temcolor-nocolor.zip)

**Comment by rhubarb-geek-nz** [comment]
> For a programmer, the correct answer is to see how it's implemented elsewhere.

Or to ask what the user requirement is.

**Comment by 237dmitry** [comment]
> demo-termcolor.py

Python 3.11.6:
```
ModuleNotFoundError: No module named 'termcolor'
```

**Comment by janko-js** [comment]
> ModuleNotFoundError: No module named 'termcolor'

On windows:

    py -m pip install termcolor

**Comment by janko-js** [comment]
Adding  -NoProfile doesn't change the bug reported in this entry. The error output is still colored and it shouldn't be. Everything else is a tangential discussion.

![demo2](https://github.com/PowerShell/PowerShell/assets/116484385/723bf62a-8c75-4a17-926a-091bce3d2fc4)

**Comment by 237dmitry** [comment]
> The error output is still colored and it shouldn't be.

![ss](https://github.com/PowerShell/PowerShell/assets/78153320/c85657c2-ffa1-417b-9f00-4300ffa6a9de)

**Comment by janko-js** [comment]
Not here:

![demo3](https://github.com/PowerShell/PowerShell/assets/116484385/306d4cf2-056b-4d23-9247-7f525f85ca52)

**Comment by janko-js** [comment]
I haven't seen in your outputs that you ever see red errors? What's the step when the red errors stop being red in your case, and how does it look before that step?

**Comment by 237dmitry** [comment]
> and how does it look before that step? I think PSReadLine is completely irrelevant for that

![ss](https://github.com/PowerShell/PowerShell/assets/78153320/01ef607b-c713-4949-9d9b-1115dffc942c)

**Comment by 237dmitry** [comment]
> I'm simply extracting

I do the same on Linux and there is not red error. What terminal emulator are you using? And I set in registry:

```
$ (Get-ItemProperty HKCU:\Console).VirtualTerminalLevel
1
```

**Comment by janko-js** [comment]
I'm opening a %windir%\system32\cmd.exe  and I haven't created any profile.ps1 myself and the only ones I see are something like in C:\Windows\System32\WindowsPowerShell\v1.0\Examples\profile.ps1 and these have only comments inside and should not be seen by the pwsh at all. And as you see adding -nop changes nothing.

I also don't have VirtualTerminalLevel anywhere in registry. Can you try to remove your entry and see what happens?

**Comment by rhubarb-geek-nz** [comment]
This gives black on white in PuTTY

![image](https://github.com/PowerShell/PowerShell/assets/94297700/d1c51508-0abf-4ab5-90ec-4149821063a2)

You can switch between black-on-white or white-on-black using the `[X] System Colours` checkbox

**Comment by janko-js** [comment]
I'm reporting here that pwsh.exe fails to behave according to NO_COLOR.

I'm not looking for workarounds using some other programs which can change behavior of all programs started through them.

**Comment by rhubarb-geek-nz** [comment]
I was looking for a way of easily testing the black text on white background scenarios, and agree that pwsh should simply be using the existing terminals colours.

**Comment by janko-js** [comment]
Why not using Windows features directly?

![cmd](https://github.com/PowerShell/PowerShell/assets/116484385/947fc7e6-58a8-471c-aded-2ecb49a2f133)

**Comment by rhubarb-geek-nz** [comment]
> Why not using Windows features directly?

I am sure there are ways, I found it the easiest way to have different profiles save under different names.

**Comment by janko-js** [comment]
>  I found it the easiest way to have different profiles

I have this also only using default features and no external programs: the shortcuts remember the settings: starting from the shortcut to cmd.exe in the taskbar, you see black text on grey. That black background was cmd.exe started from the start menu. 
![cmd2](https://github.com/PowerShell/PowerShell/assets/116484385/ad6c9a76-3bfb-48a2-afa2-0dc71cebfbdd)


But that really doesn't matter for this issue. The issue is still:

**The pwsh.exe fails to behave according to NO_COLOR.**

**Comment by 237dmitry** [comment]
You can switch between white and black background/foreground:

only background and foreground:
```
"`e]4;0;#ffffff;7;#000000`a" 
```
all 16 colors:
```
$esc = (1..15 | % { '{0};{1}' -f $_,'#000' }) -join ';'
"`e]4;0;#fff;$esc`a"
$PSStyle.OutputRendering = "PlainText"
```
This does not affect PSReadline. It has to be configured separately.

**Comment by jhoneill** [comment]
Just seeing the initial
```
OS                             Microsoft Windows 6.3.9600
```
Is this a facet of the way PowerShell behaves on older console hosts. ? 

<img width="448" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/e8b1a528-c3e7-47d0-bb78-ae8843d737d8">

As you can see on Windows 11 with Windows Terminal,  It works as expected. BUT PSReadline (IIRC) does not use ansi sequences 
<img width="502" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/9f405e1c-f83a-42bb-b33a-5f93d9bf1dd2">
 In the no_color instance writing the ansi blue sequence has no effect, but unsurprisingly Write-host goes around it, and I believe that's the method used by PSReadline.

**Comment by 237dmitry** [comment]
> BUT PSReadline (IIRC) does not use ansi sequences

If you type `Get-PSReadlineOption` you can see that it uses them. This is by default:

```
CommandColor                           : "`e[93m"
CommentColor                           : "`e[32m"
ContinuationPromptColor                : "`e[37m"
DefaultTokenColor                      : "`e[37m"
EmphasisColor                          : "`e[96m"
ErrorColor                             : "`e[91m"
InlinePredictionColor                  : "`e[97;2;3m"
KeywordColor                           : "`e[92m"
ListPredictionColor                    : "`e[33m"
ListPredictionSelectedColor            : "`e[48;5;238m"
ListPredictionTooltipColor             : "`e[97;2;3m"
MemberColor                            : "`e[37m"
NumberColor                            : "`e[97m"
OperatorColor                          : "`e[90m"
ParameterColor                         : "`e[90m"
SelectionColor                         : "`e[30;47m"
StringColor                            : "`e[36m"
TypeColor                              : "`e[37m"
VariableColor                          : "`e[92m"
```

**Comment by rhubarb-geek-nz** [comment]
Unfortunately the colours can't be defined as $Null to say do not set the colour. However the goal is not to have the colours configured to achieve the effect, the goal is to have all colours ignored when NO_COLOR is set.

**Comment by rhubarb-geek-nz** [comment]
> > the goal is to have all colours ignored when NO_COLOR is set
> 
> Yes. That's why [#3918](https://github.com/PowerShell/PSReadLine/issues/3918)

Yes, but what we missed was NO_COLOR affects the output channel, but not the error channel in pwsh.

**Comment by rhubarb-geek-nz** [comment]
> Everything works for me, stderr is not red, look at screenshots above, why it doesn’t work for you, I don’t know. Maybe it depends on the version of Windows.

Yes, just tested on Linux

```
$ NO_COLOR=true pwsh -c "this-does-not-exist"
this-does-not-exist: The term 'this-does-not-exist' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

and that did not print any colours. But if removed the NO_COLOR=true then it was all red.

**Comment by mklement0** [comment]
@janko-js:

* `Microsoft Windows 6.3.9600` suggests you're running Windows 8.1, which is no longer supported (since v7.2, the minimum version has been Windows 10) - see https://learn.microsoft.com/en-us/powershell/scripting/install/powershell-support-lifecycle?view=powershell-7.4#supported-platforms
  * While the MSI installer in 7.4 includes a minimum version check, unfortunately it apparently thinks Windows 8.1 is still supported:
  * https://github.com/PowerShell/PowerShell/blob/4f02a89906a92f684ff4511ae74958f602edf800/assets/wix/Product.wxs#L136-L141
 
* Like others, I don't see the problem (on my W11 22H2 machine). If you can reproduce the problem on a still-supported OS version, please let us know.

* A few general points:

  *  The _interactive_ (input) shell experience is _not_ covered by the `NO_COLOR` and `TERM` behavioral contract, if you will - only _output_.
      * The command-line editing experience and its coloring is provided by a separate module, `PSReadLine`, and possibly adding support for a monochrome mode for it too is the subject of the following _feature request_ in the latter's repo: https://github.com/PowerShell/PSReadLine/issues/3918

  * In repro steps involving the PowerShell CLI, it is advisable to use `-NoProfile`, so as to rule out potential side effects from profiles.

  * The `VirtualTerminalLevel` registry setting is unrelated to the issue at hand: it tells `conhost.exe` whether regular console windows should render VT (ANSI) sequences as such when emitted by programs. However, PowerShell manages this setting dynamically: for output _from PowerShell commands_, it _unconditionally_ enables this support, and for output _from external programs_, it respects the registry setting.
    * The dynamic switching between support and non-support is currently broken in pipelines involving an external program followed by a PowerShell command that produces colored output (Windows Terminal is _not affected), e.g. `cmd /c echo hi | Select-String h` - see #20898

**Comment by jhoneill** [comment]
> > BUT PSReadline (IIRC) does not use ansi sequences
> 
> If you type `Get-PSReadlineOption` you can see that it uses them. This is by default:
> 
> ```
> CommandColor                           : "`e[93m"
> CommentColor                           : "`e[32m"
> ```
Yes but I have something in my memory which says that's not what actually goes to the console. PSR knows how to translate "blue" to an ansi sequence, and I wish I could find where I got the idea that it does the equivalent of a `write-host -foreground` and not `write-output "ANSI_seq"` 
But that is what this the screen shot below suggests
<img width="501" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/cc1404f0-b14d-4ea8-bdc3-e5d5caf23568">
 Note the colour goes from the battery-life bars in my prompt as well, it is only PSR which seems able to emit colour.

**Comment by 237dmitry** [comment]
> what actually goes to the console

The "blue" is 94m ))

![ss](https://github.com/PowerShell/PowerShell/assets/78153320/8f62ccfd-bac1-4656-881a-5035d6c27264)

You can test color palette:

```powershell
0..15 | % { Write-Host $([consoleColor] $_) -f $_ }
```

**Comment by jhoneill** [comment]
> > what actually goes to the console
> 
> The "blue" is 94m ))
> 
> 
> You can test color palette:
> 

Yes. But  `Write-host -foreground Blue ` and   `write-host "`e[94mblue"` work differently. `No_color` renders ANSI escape sequences ineffective but doesn't stop  `-Foreground` working. 

<img width="470" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/0f171da9-b6af-4677-95ec-f2a9c6ae81ee">

I do not **know** for 100% certain what happens inside PSReadline. But the fact its colours are not disabled when colours of everything else using ANSI **is** would *suggest* it is translating the ANSI it is given to console commands, but I can't find where I got the idea that is what is happening.

**Comment by 237dmitry** [comment]
This behavior is strange. `$env:NO_COLOR = 1` sets `$PSStyle.OutputRendering = "PlainText"`, but this has no effect on `Write-Host` with `System.ConsoleColor` enum in parameters.

But if I entirely set mono for terminal then Write-Host is also monochrome:

```powershell
# mono.ps1


param ( [Parameter()] [string] $color = "black" )

if ($color -eq "white") { $bg, $fg = "#ccc", "#000" }
else { $bg, $fg = "#111", "#ddd" }

[console]::Write("`e]12;$fg`a")

$PSStyle.OutputRendering = "PlainText"
Set-PSReadlineOption -Colors @{
    Command = "`e[37m"
    Comment = "`e[37m"
    ContinuationPrompt  = "`e[37m"
    Default = "`e[37m"
    Emphasis = "`e[37m"
    Error = "`e[37m"
    InlinePrediction  = "`e[37m"
    Keyword = "`e[37m"
    ListPrediction = "`e[37m"
    ListPredictionSelected  = "`e[37m"
    ListPredictionTooltip  = "`e[37m"
    Member = "`e[37m"
    Number  = "`e[37m"
    Operator  = "`e[37m"
    Parameter  = "`e[37m"
    Selection =  "`e[7m"
    String  = "`e[37m"
    Type =  "`e[37m"
    Variable  = "`e[37m" 
}

$col = (1..15 | Foreach-Object { '{0};{1}' -f $_, $fg }) -join ';'
[console]::Write("`e]4;0;$bg;$col`a")
```
Works in conhost and it can work in terminals without living config reloading:
```
PS > mono white  # or black
```

**Comment by mklement0** [comment]
@237dmitry, the `Write-Host` behavior seems to be an (actual) bug - or at least a baffling cross-platform inconsistency; see:

* #21168

**Comment by jhoneill** [comment]
> This behavior is strange. `$env:NO_COLOR = 1` sets `$PSStyle.OutputRendering = "PlainText"`, but this has no effect on `Write-Host` with `System.ConsoleColor` enum in parameters.
> 
> But if I entirely set mono for terminal then Write-Host is also monochrome:
> 
> ```powershell
> # mono.ps1
> 
> 
> param ( [Parameter()] [string] $color = "black" )
> 
> if ($color -eq "white") { $bg, $fg = "#ccc", "#000" }
> else { $bg, $fg = "#111", "#ddd" }
> 
> [console]::Write("`e]12;$fg`a")
> 
> $PSStyle.OutputRendering = "PlainText"
> Set-PSReadlineOption -Colors @{
>     Command = "`e[37m"
>     Comment = "`e[37m"
>     ContinuationPrompt  = "`e[37m"
>     Default = "`e[37m"
>     Emphasis = "`e[37m"
>     Error = "`e[37m"
>     InlinePrediction  = "`e[37m"
>     Keyword = "`e[37m"
>     ListPrediction = "`e[37m"
>     ListPredictionSelected  = "`e[37m"
>     ListPredictionTooltip  = "`e[37m"
>     Member = "`e[37m"
>     Number  = "`e[37m"
>     Operator  = "`e[37m"
>     Parameter  = "`e[37m"
>     Selection =  "`e[7m"
>     String  = "`e[37m"
>     Type =  "`e[37m"
>     Variable  = "`e[37m" 
> }
> 
> $col = (1..15 | Foreach-Object { '{0};{1}' -f $_, $fg }) -join ';'
> [console]::Write("`e]4;0;$bg;$col`a")
> ```
> 
> Works in conhost and it can work in terminals without living config reloading:
> 
> ```
> PS > mono white  # or black
> ```

`Write-host` is not selecting (e.g) RGB of 0,0,255 for blue it's selecting terminal colour 9 
What your code does is set all the terminal colours to the same thing
<img width="645" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/136a1b3b-1797-4cec-86f9-5bafa10c8e29">

Note also (1)
<img width="663" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/ea3f4c19-dcdb-4fe8-b834-00d36d4505c5">
`Write-host` strips ansi when No_color is true, but `-Foreground`  is not affected. 
(2) What in one place is called blue is either "BrightBlue" or "DarkBlue" in the the other
(3) These changes affect the console even after the process which applied them exits.

**Comment by rhubarb-geek-nz** [comment]
> `Write-host` strips ansi when No_color is true, but `-Foreground` is not affected.

Rather than "stripping ansi", I suggest code should not be writing ANSI colour escape sequences to the output in the first place.

**Comment by 237dmitry** [comment]
> What your code does is set all the terminal colours to the same thing

Yes. What does it do to reset color palette to two colors (background and foreground)? Monochrome is two colors 37m and 40m and their values are taken from the terminal settings, where the 1st is background and the 8tn is foreground.

**Comment by rhubarb-geek-nz** [comment]
> Monochrome is two colors 37m and 40m and their values are taken from the terminal settings, where the 1st is background and the 8tn is foreground.

If the terminal is monochrome terminal and does not handle ANSI colour escape sequences and it will just print out the text the best it can, so your output would be littered with the literals "37m" and "40m" in the content.

**Comment by mklement0** [comment]
@rhubarb-geek-nz, regarding `Write-Host`  stripping vs. not writing VT escape sequences:

* I agree that it should not write them to begin with when `-ForegroundColor` / `-BackgroundColor` are used - that's the subject of #21168; the fact that it _always_ writes them _on Windows_ (whereas  on Unix only with `NO_COLOR` it does not)  is probably due to the fact that the legacy [`WriteConsoleOutput`](https://learn.microsoft.com/en-us/windows/console/writeconsoleoutput) WinAPI function is used there, whose color support is unrelated to and predates VT support.

* Separately, active escape-sequence stripping is performed when `$PSStyle.OutputRendering = 'PlainText'` is in effect, as follows:
  * When _PowerShell_ performs to-terminal output , as opposed to an external program call's output being _passed through_ to the terminal.
  * Intra-session, this only affects the _display_ of data, but an outside caller actually receives the _stripped_ text.

Here's a Unix example; run from `bash`:

```sh
# "green" prints in green, because the external printf utility prints directly to the terminal
# (and itself doesn't honor NO_COLOR)
NO_COLOR=1 pwsh -nop -c "printf 'It ain''t easy being \033[32mgreen\033[m.\n'"

# "green" does NOT print in green, because the use of (...) means that PowerShell captures and relays
# the output, and performs active VT-sequence stripping due to NO_COLOR -> $PSStyle.OutputRendering = 'PlainText'
NO_COLOR=1 pwsh -nop -c "printf 'It ain''t easy being \033[32mgreen\033[m.\n'"
```

**Comment by 237dmitry** [comment]
> Here's a Unix example; run from bash:

This has to be colored or not?

![ss](https://github.com/PowerShell/PowerShell/assets/78153320/45538e7c-f3e4-4512-beed-fb73dfc1c55c)

**Comment by rhubarb-geek-nz** [comment]
If you explicitly print any escape sequence then that is what should be output because that is what you said to do.

With NO_COLOR there should be no need for any filtering, it is up to the code to abide by the request to not output any colour.

**Comment by mklement0** [comment]
My bad, @237dmitry - I forgot to add the `(...)` enclosure, which triggers PowerShell's VT-sequence stripping  (I've updated my comment):

```sh
# From Bash
NO_COLOR=1 pwsh -nop -c "(printf 'It ain''t easy being \033[32mgreen\033[m.\n')"
```

Now the word "green" should NOT be colored.

---

@rhubarb-geek-nz 

> If you explicitly print any escape sequence then that is what should be output 

I don't disagree, but - unlike the buggy, inconsistent `Write-Host -Fore/BackgroundColor` behavior - this stripping is by intentional design.

That said, you could argue that _outside callers_, specifically, then see _actively modified_ data is a buggy (possibly accidental implementation) aspect of this design.
(Intra-session, the problem doesn't arise: the data itself isn't modified during further processing, only the _display_ is impacted).

In other words: it may worth opening a separate issue to question the _CLI_ behavior when stripping is in effect.

**Comment by 237dmitry** [comment]
> If the terminal is monochrome terminal and does not handle ANSI colour escape sequences

Try to see rgb value of background (for example):

```
"`e]4;0;?`a"
```
All colors are escape sequences. Why they do not handle? I think they are write directly into the host.

**Comment by rhubarb-geek-nz** [comment]
> All colors are escape sequences. Why do they not handle? I think they write directly into the host.

All colours are escape sequences, and monochrome terminals _do not deal with colours_.

**Comment by 237dmitry** [comment]
> and monochrome terminals do not deal with colours.

Where does your yellow-black terminal get the colors?

**Comment by rhubarb-geek-nz** [comment]
> Where does your yellow-black terminal get the colors?

I don't know and I don't care because that is internal to the terminal implementation. What I want is my terminals to work with their existing themes.

Years ago I wrote a nice simple terminal, [WishStream](https://apps.microsoft.com/detail/9NBLGGH679VT), this was Windows Mobile/Windows 8 Store app. The foreground and background colours were defined by the user through the XAML Theme, if they wanted it dark it would be light on dark, if they wanted light, they got dark on light. 

As it was a monochrome terminal it implemented the VT100 escape sequences, which do not include colour. So this terminal ignored all colour escape sequences because they were meaningless, and as it set TERM=vt100, no application should have been sending any colour escape sequences. So vi etc all worked happily because they abided by the terminfo regarding what a vt100 terminal could do.

**Comment by 237dmitry** [comment]
> What I want is my terminals to work with their existing themes.

Сonfigure pwsh and terminal yourself. This is the easiest way in the current situation. Or wait for the PSReadline developers to listen to our wishes.

**Comment by rhubarb-geek-nz** [comment]
> Сonfigure pwsh and terminal yourself. This is the easiest way in the current situation. Or wait for the PSReadline developers to listen to our wishes.

The other option is to not use PowerShell. It is a silly situation where, out of the box, with no configuration, PowerShell is unusable on a Mac with the default terminal theme of black on white because it gives you yellow on white. I have to open a new session explicitly different to use PowerShell.

**Comment by 237dmitry** [comment]
I could not launch bash in monochrome mode, all of `TERM=mono|vt100`, `NO_COLOR=1|true` do not work at all. only for one-line syntax and not for all applications and utilities. Even in TTY (TERM=linux).

**Comment by 237dmitry** [comment]
If I launched bash:
```
bash --noprofile --norc
```
It still started with full set of configurations from `~/.profile` and `~/.bashrc`

**Comment by jhoneill** [comment]
> > `Write-host` strips ansi when No_color is true, but `-Foreground` is not affected.
> 
> Rather than "stripping ansi", I suggest code should not be writing ANSI colour escape sequences to the output in the first place when NO_COLOR is present or OutputRendering = PlainText.

I would suggest that authors need to think REALLY hard about use of any kind of colour, and a lot harder than any actually do.
I think the chance of getting authors to write `if  no_color` around their not terrible well thought out colours isn't great. 

If a script that doesn't check the users preference uses`Write-host  "$($PSStyle.Foreground.Yellow) Something $($PSStyle.Reset)"`  in place of the frankly easier and longer established `Write-host  "Something" -Foreground  yellow`   then PowerShell will apply NO_COLOR as the user wishes.  

>  What I want is my terminals to work with their existing themes
There is always the scope to create  terminal themes which become nonsensical . _Terminal_ themes are a set of "slots" which represent colours by convention, and we can set exactly what R,G,B values are used for "Red-on-this-terminal" VS Code's default red and / Windows Terminal default Red are not automatically the same or something simple like 255,0,0.  Windows PowerShell re-defined Dark Magenta in the default terminal settings for PowerShell.exe so that it isn't any kind of magenta but a dark blue and then set the background not to black but dark magenta.   

Absolutely any assumption that an _application_ theme makes about the terminal has in each slot can be wrong footed:
what you or I set for winterm / vs code - because of taste or address an eyesight issue we have - can result in PSReadline displaying unhelpfully. 

I think there is an argument that at start-up PSreadline should look at No_color and if found set all it's colours to the default foreground colour. That's a different branch from this though.

**Comment by SteveL-MSFT** [comment]
I'm a bit lost in this thread.  The original problem doesn't repro for me:

```powershell
$env:NO_COLOR=1
pwsh -noprofile -command '$psversiontable;get-it'
```

![image](https://github.com/PowerShell/PowerShell/assets/11859881/8d005007-6748-4967-8617-ac8479d0412b)

**Comment by mklement0** [comment]
@SteveL-MSFT, I don't think there's anything actionable _here_, as I've tried to express in my [previous summary](https://github.com/PowerShell/PowerShell/issues/21160#issuecomment-1917955913):

* We have no repro for the original problem, except on a _no longer supported Windows version, 8.1_, which makes it a moot point. (Also, the OP appears to have deleted their account.)

* The conceptually related, but separate question of _PSReadLine_'s behavior with respect to coloring of _interactive user input_ at the command prompt should be discussed in:
  * https://github.com/PowerShell/PSReadLine/issues/3918

**Comment by bitbound** [comment]
Landed here while searching the same issue.  My use case is that I'm starting pwsh with redirected stdin, stdout, and stderr.  With `NO_COLOR` environment variable set, I'd expect there to be no color escape sequences in neither stdout nor stderr.

Maybe I'm wrong to assume this, but it seems like a logical conclusion based on the docs here:  https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_ansi_terminals?view=powershell-7.4#disabling-ansi-output

Here's a quick repro:

```C#
using System.Diagnostics;

var psi = new ProcessStartInfo
{
    FileName = "pwsh.exe",
    RedirectStandardError = true,
    RedirectStandardOutput = true,
    RedirectStandardInput = true,
};
psi.Environment.Add("NO_COLOR", "1");

var process = Process.Start(psi)!;
process.OutputDataReceived += (sender, e) =>
{
    Console.WriteLine(e.Data);
};
process.ErrorDataReceived += (sender, e) =>
{
    Console.WriteLine(e.Data);
};

process.BeginOutputReadLine();
process.BeginErrorReadLine();

await process.StandardInput.WriteLineAsync("$env:NO_COLOR;$psversiontable;get-it;exit;");
await process.WaitForExitAsync();
```

![image](https://github.com/PowerShell/PowerShell/assets/20995508/c3af8d49-4c9c-4837-927f-2a1b48854a3d)

![image](https://github.com/PowerShell/PowerShell/assets/20995508/ec854d5b-a979-414f-aa54-61f788eaa5c1)

**Comment by akzov** [comment]
I set NO_COLOR:1 as a machine environment variable but still get color codes in Azure Automation logs

**Comment by Zorbn** [comment]
This also happens when redirecting stdin, stdout, stderr with Win32. Setting the __SuppressAnsiEscapeSequences environment variable makes it work as expected.

**Comment by mklement0** [comment]
Thanks for mentioning the `__SuppressAnsiEscapeSequences` environment-variable workaround, but it's not a real solution, given the comment made by a team member in https://github.com/PowerShell/PowerShell/issues/19188#issuecomment-1601562724:

> we're a little concerned about your use of `__SuppressAnsiEscapeSequences` as it is a `__*` variable which are generally (and traditionally) internal variables. (which is one of the reasons it's not documented)

Another related issue:

* #19961

As noted in the linked issue, it seems that the bug surfaces when redirecting _stderr_ output, specifically, and I've posted a simple repro [there](https://github.com/PowerShell/PowerShell/issues/19961#issuecomment-2356355977).

Since redirecting stderr is also necessary to make the repro in _this_ issue's initial post produce the symptom, I suggest closing this issue in favor of the aforementioned #19961.
